node/src/actors/chain_manager/handlers.rs

```diff
@@ -306,118 +302,234 @@ impl Handler<AddBlocks> for ChainManager {
+                                    log_sync_progress(&sync_target, &blocks2, num_processed_blocks, 2);
+
+                                    // Update ARS if there were no blocks right before the epoch during
+                                    // which we should construct the target superblock
+                                    let candidate_superblock_epoch = (act.current_epoch.unwrap() / superblock_period) * superblock_period;
+
+                                    // We need to persist blocks in order to be able to construct the
+                                    // superblock
+                                    act.persist_blocks_batch(ctx, blocks2);
+                                    let to_be_stored =
+                                        act.chain_state.data_request_pool.finished_data_requests();
+                                    act.persist_data_requests(ctx, to_be_stored);
+
+                                    log::info!("Block sync target achieved, go to WaitingConsensus state");
+                                    // Target achieved, go back to state 1
+                                    act.sm_state = StateMachine::WaitingConsensus;
+
+                                    // We must construct the second superblock in order to be able
+                                    // to validate the votes for this superblock later
+                                    log::debug!("Will construct the second superblock during synchronization. Superblock index: {} Epoch {}", sync_target.superblock.checkpoint + 1, candidate_superblock_epoch);
+
+                                    actix::fut::ok(candidate_superblock_epoch)
+                                }
+                            })
+                            .and_then(move |candidate_superblock_epoch, act, ctx| {
+                                act.build_and_vote_candidate_superblock(ctx, candidate_superblock_epoch)
+                            })
+                            .and_then(move |_, act, ctx| {
+                                // Process remaining blocks
+                                let (batch_succeeded, num_processed_blocks) = act.process_blocks_batch(ctx, &sync_target, &blocks3);
+                                if !batch_succeeded {
+                                    log::error!("3 Received invalid blocks batch");
+                                    act.sm_state = StateMachine::WaitingConsensus;
+                                    act.sync_waiting_for_add_blocks_since = None;
+
+                                    return actix::fut::err(());
+                                }
+                                log_sync_progress(&sync_target, &blocks3, num_processed_blocks, 3);
+
+                                log::info!("Block sync target achieved, go to WaitingConsensus state");
+                                // Target achieved, go back to state 1
+                                act.sm_state = StateMachine::WaitingConsensus;
```

Comment:
In lines 457-459 we already do the same, it is necessary?

Reply:
I think so :thinking: